README
	-*- Indented-text -*-
$Id: README 457 2014-03-09 14:44:24Z hayashi $

Term::ReadLine::Gnu --- GNU Readline Library Wrapper Module

	Copyright (c) 2014 Hiroo Hayashi.  All rights reserved.

	This program is free software; you can redistribute it and/or
	modify it under the same terms as Perl itself.

Description:

	Term::ReadLine::Gnu (TRG) is an implementation of the
	interface to the GNU Readline Library.  This module gives you
	input line editing facility, input history management
	facility, word completion facility, etc.  It uses the real GNU
	Readline Library and has the interface with the almost all
	variables and functions which are documented in the GNU
	Readline/History Library.  So you can program your custom
	editing function, your custom completion function, and so on
	with Perl.  TRG may be useful for a C programmer to prototype
	a program which uses the GNU Readline Library.

	TRG is upper compatible with Term::ReadLine included in Perl
	distribution.  Term::ReadLine uses TRG automatically when TRG
	is available.  You can enjoy full line editing feature with
	Perl debugger which use Term::ReadLine with no patch.

	Ilya Zakharevich distributes his implementation,
	Term::ReadLine::Perl, which bases on Jeffrey Friedl's
	readline.pl.  His module works very well, and is easy to
	install because it is written by only Perl.  I am trying to
	make my module compatible with his.  He gives useful advises
	for me.  Unfortunately readline.pl simulated old GNU Readline
	library before TRG was born.  For example, it was not 8 bit
	clean and it warns to the variables in ~/.inputrc which it did
	not know yet.  We Japanese usually use 8 bit characters, so
	this was bad feature for me.  I could make a patch for these
	problems but I had interest with C interface facility and
	dynamic loading facility of Perl, so I thought it was a good
	chance for me to study them.  Then I made this module instead
	of fixing his module.

Prerequisites:
	You must have Perl 5.7 or later.  If you have to use old Perl
	for some reason, use Term::ReadLine::Gnu 1.09.  (I recommend
	you to use newer Perl.)

	You must have GNU Readline Library Version 2.1 or later.  See
	INSTALL for more detail.

	A report said GNU Readline Library might not work with perl with
	sfio.  Since I do not have sfio library, I am not sure.

How to build/install:
	See INSTALL.

Bugs:
	There may be many bugs in both programs and documents.
	Comments and bug reports are very welcome. Send me a E-Mail or
	open a ticket on the bug tracker at rt.cpan.org;

	http://rt.cpan.org/Dist/Display.html?Queue=Term-ReadLine-Gnu

Author:
	Hiroo Hayashi <hiroo.hayashi@computer.org>

Project Home Page
	http://sourceforge.net/projects/perl-trg/

Revision History:
	See Changes.


Changes
	-*- Indented-text -*-
	$Id: Changes 476 2014-12-13 03:56:06Z hayashi $

1.25  2014-12-13
	- fix to call rl_initialize() after I/O stream
	  setting. [rt.cpan.org #96569]
	- t/readline.t: comment-out stty-command calls. 
	- t/history.t, t/readline.t: print out the corresponding
          section names in the manuals of the libraries.
	- eg/perlsh: use 'IO' instead of 'FILEHANDLE' which Perl 5.20
	  warns. [rt.cpan.org #100883]
	- Fix copyright notices to use the year of first publication.

1.24  2014-03-23
	- fix not to make the PerlIO layer empty for Perl 5.8.9 or
          before. [rt.cpan.org #59832]
	- update Pod document assuming the use of Pod::Simple::HTMLBatch.
	- t/00checkver.t: not to use 'display-readline-version' and "use
	  Test;"
	- INSTALL: add "(Install) Using Package" section.

1.23  2014-03-20
	- add description of the pager command bug in Perl debugger
	  which causes segementation fault [rt.perl.org #121456]
	- fix not to clobber binmode layer (utf-8) on filehandles [rt.cpan.org #59832]
	- support UTF-8 input as compatible with Term::ReadLine
	  - Now Term::ReadLine::Gnu requires perl 5.8.0 or above
	- define Term::ReadLine::Gnu::ornaments method [rt.cpan.org #93614]
	- Makefile.PL: Fix the URL of the bug-tracker [rt.cpan.org #93680]
	- eg/perlsh: POD syntax fix. [rt.cpan.org #93895]
	- README: add note for bug-tracker
	- INSTALL: update "1.3 Trouble Shooting"

1.22  2014-03-05
	- make Kwalitee <http://cpants.cpanauthors.org/> friendly
	  - add MakeMaker options for META.yml
	  - add Changelog
	  - fix POD errors
	  - make .pm versions consistent
	  - use "use warnings"
	- Makefile.PL dies on the CPAN Test for OpenBSD temporarily
          (hopefully).

1.21  2014-03-02
	- readline-6.3 support (Gnu Readline Library 6.3 requires this
	  release or newers.)
	    new function
		rl_clear_history
	    new variable
		rl_signal_event_hook
		rl_input_available_hook
		rl_executing_key
		rl_executing_keyseq
		rl_key_sequence_length
		rl_change_environment
		rl_filename_stat_hook
	- readline-6.1 support
	    new function
		rl_free_keymap (enabled)
	    new variable
		rl_filename_rewrite_hook (now implemented)
	- make the following variables read-only.: rl_executing_macro,
          history_length, rl_readline_state, rl_explicit_arg,
	  rl_numeric_arg, rl_editing_mode
	- Improve backward compatiblity variable and function definitions.
	- Fix Term::ReadLine::Gnu::XS::_trp_completion_function not to
	  perform a case-sensitive filter. [rt.cpan.org #72378]
	- fix Gnu.xs:rl_initialize to be compiled with Perl 5.8.5 or
          before. [rt.cpan.org #61626]
	- update Pod documents

	- t/readline.t, t/history.t: add variable access tests.
	- t/readline.t: consider .svn directory on filename
          completion test and to set the standard tty setting before
          re_initialize() being called.
	- eg/perlsh makes use of the deferred signal (safe signal)
          feature.

	- Maefile.PL checks whether the xmalloc exported by readline
	  is indeed called 'xmalloc' or whether it has been renamed to
	  '_rl_malloc'. [rt.cpan.org #65973]
	- Makefile.PL: add -DTRL_READLINE_VERSION to show GNU Readline
          Library version in CPAN Tester Reports.
	- t/00checkver.t: let messages go to stdout.

	- README: add project home page URL and remove out-of-date
          descriptions.
	- INSTALL: add a section "Install on Mac OS X"

1.20  2010-05-02
	- Make a copy of the environment variable array to stop
          segmentation faults on some systems (ex. FreeBSD)
	- t/readline.t skip the test of rl_readline_version for GNU
	  Readline Library 6.1 which may return a wrong value [rt.cpan.org #54977]
	- readline-6.1 support
	    new function
		rl_free_keymap (disabled due to readline-6.1 bug)
	    new variable
		rl_filename_rewrite_hook (not implemented yet)

1.19  2009-03-21
	- make sure the outstream fd inside the readline library is in
          sync [rt.cpan.org #16440]
	- fixes to be more CPAN Testers friendly
	  - t/readline.t does not use visible bell for "make test"
	  - Makefile.PL requires perl 5.7.0 or later.  (Use
	    Term::ReadLine::Gnu-1.09 for older Perl.)
	  - Makefile.PL exits 0 when /dev/tty cannot be opened.
	- INSTALL
	    add description of the -DPERL_USE_SAFE_PUTENV issue.  If
	    you encounter a segmentation fault, read it. [rt.cpan.org #37194]

1.18  2009-02-27
	- better error checking of Makefile.PL for automatic testing
	  on non-supported platfrom (for example one without GNU
	  Readline Library).
	- readline-6.0 support
	    new function
		rl_save_state (not supported yet)
		rl_restore_state (not supported yet)
		rl_echo_signal_char
	    new variable
		rl_display_prompt
		rl_sort_completion_matches
		rl_completion_invoking_key
	    t/history.t does not fail by hist_expand() which is fixed
	    on readline-6.0

1.17  2008-02-07
	- Unnecessary PerlIO_releaseFILE() calls are removed.  This
          fix prevents `make test' from failing on Perl 5.10.
	- Makefile.PL now fails with the EditLine Library on MacOS X.
          Use the GNU Readline Library. [rt.cpan.org #28523]
	- tested with readline-5.2 (which has no new feature to be
          supported).

1.16  2006-04-02
	- fix Makefile.PL to work with the latest Cygwin.
         (Use 1.15 for old Cygwin.)
	- readline-5.1 support
	    new function
		variable_value
		reset_screen_size
	    new variable
		prefer_env_winsize

1.15  2004-10-17
	- readline-5.0 support
	    new function
		bind_key_if_unbound
		bind_keyseq
		bind_keyseq_if_unbound
		tty_unset_default_bindings
		add_history_time
		history_get_time
	    new variable
		history_write_timestamps
		completion_quote_character
		completion_suppress_quote
		completion_found_quote
		completion_word_break_hook
	- double IO stream close bug fix (more use of PerlIO) [rt.cpan.org #7672]
	- warning on 'use Term::ReadLine::Gnu;'.

1.14  2003-03-16
	- kludge not to cause segmentation fault on Perl 5.8.0
          w/PerlIO and FileHandle (ex. CPAN.pm)

Makefile.PL
#
#	Makefile.PL for Term::ReadLine::Gnu
#
#	$Id: Makefile.PL 475 2014-12-13 03:20:00Z hayashi $
#
#	Copyright (c) 1996 Hiroo Hayashi.  All rights reserved.
#		<hiroo.hayashi@computer.org>
#
#	This program is free software; you can redistribute it and/or
#	modify it under the same terms as Perl itself.
#
#	OS/2 support was contributed by Ilya Zakharevich.
#		<ilya@math.ohio-state.edu>
#
# Usage: perl Makefile.PL [--prefix=...] [--includedir=...] [--libdir=...]
#		[OPTIMIZE=...]
#
#	Read INSTALL for more details.
########################################################################
use strict;
use ExtUtils::MakeMaker;
use Config;
use Getopt::Long;
use 5.008; use 5.8.0;
my ($defs, $libs, $lddflags, $RLLIB, $RLINC);

# exit 0 before creating the Makefile to be CPAN Testers friendly
# see http://wiki.cpantester.org/wiki/CPANAuthorNotes
my $err = $ENV{AUTOMATED_TESTING} ? 0 : 1;
# test /dev/tty for CPAN Testers
if ($ENV{AUTOMATED_TESTING} && !open(SESAMI, '/dev/tty')) {
    warn "cannot open /dev/tty\n";
    exit 0;
}
# I could not reach a tester, "Chris Williams (BINGOS)".
if ($ENV{AUTOMATED_TESTING} && ($Config{osname} eq 'openbsd')) {
    die "OS unsupported\nPlease contact the author.\n";
}

$defs = ($Config{strings} =~ m|/string.h$|) ? '-DHAVE_STRING_H' : '';

# Parse command line to specify paths for the GNU Readline Library
{
    my ($prefix, $libdir, $incdir);
    GetOptions("prefix=s" => \$prefix,
	       "libdir=s" => \$libdir,
	       "includedir=s" => \$incdir);
    $RLLIB = defined $libdir
	? "-L$libdir" : (defined $prefix ? "-L$prefix/lib" : '');
    $RLINC = defined $incdir
	? "-I$incdir" : (defined $prefix ? "-I$prefix/include" : '');
}

if ($Config{osname} eq 'os2') {
    # Check ftp://ftp.math.ohio-state.edu/pub/users/ilya/os2/
    $libs = '-lreadline_import';
    $defs .= ' -DOS2_USEDLL';
    $lddflags = '';
} else {
    # Search libtermcap, libncurses, or libcurses in this order.
    # I emulate the behavior of the configure script for bash, and don't
    # know why AIX prefers curses.
    # libtermcap.a on HPUX cannot be used for dynamically linked binary.
    # Old Cygwin may require setting false (0).
    my $PREFER_CURSES = $Config{osname} eq 'aix' || $Config{osname} eq 'hpux'
			|| $Config{osname} eq 'cygwin';
    my $TERMCAP_LIB = (! $PREFER_CURSES && &search_lib('-ltermcap'))
	|| &search_lib('-lncurses')
	    || &search_lib('-lcurses');

    unless ($TERMCAP_LIB) {
	warn "Could not find neither libtermcap.a, libncurses.a, or libcurses.\n";
	exit $err;
    } 
 
    $libs = "-lreadline $TERMCAP_LIB";
    # Latest Perl in FreeBSD does not need this hack. (Dec.2002)
    $libs .= ' -lcrypt' if ($Config{osname} =~ /freebsd/i);
    $lddflags = '';
    # If you are using old Cygwin, enable the following line.
    #$lddflags = ($Config{osname} =~ /cygwin/i) ? '-static' : '';
}

# Check version of GNU Readline Library (for version 4.2 and before)
{
    my ($rlmajorver, $rlminorver) =
	check_readline_version($RLINC, $RLLIB, $defs, $lddflags, $libs);

    if ($rlmajorver < 4 || $rlmajorver == 4 && $rlminorver <= 2) {
	$defs .= " -DRL_READLINE_VERSION=" .
	    sprintf("0x%02x%02x", $rlmajorver, $rlminorver);
	$defs .= " -DRL_VERSION_MAJOR=$rlmajorver";
	$defs .= " -DRL_VERSION_MINOR=$rlminorver";
    }
    # to show GNU Readline Library version in CPAN Tester Reports
    $defs .= " -DTRG_READLINE_VERSION=" .
	sprintf("0x%02x%02x", $rlmajorver, $rlminorver);
}

# Check whether xmalloc has been renamed to _rl_malloc (Gentoo Linux)
if (my $extra_defs = guess_malloc_names($RLINC, $RLLIB, $defs, $lddflags, $libs)) {
    $defs .= " $extra_defs";
}

# generate a Makefile
WriteMakefile
    (
     ABSTRACT_FROM => 'Gnu.pm',
     AUTHOR	  => 'Hiroo HAYASHI <hiroo.hayashi@computer.org>',
     NAME	  => 'Term::ReadLine::Gnu',
     LICENSE	  => 'perl',
     EXE_FILES	  => [ 'eg/perlsh' ],
     META_MERGE	  => {
	 resources => {
	     license => 'http://dev.perl.org/licenses/',
	     homepage => 'http://sourceforge.net/projects/perl-trg/',
	     bugtracker => 'http://rt.cpan.org/Public/Dist/Display.html?Name=Term-ReadLine-Gnu',
	     repository => 'https://sourceforge.net/p/perl-trg/code/HEAD/tree/',
	 },
     },
     VERSION_FROM => 'Gnu.pm',
     MIN_PERL_VERSION => '5.8.0',
     LIBS	  => [ "$RLLIB $libs" ],
     LDDLFLAGS	  => "$RLLIB $Config{lddlflags}",
     dynamic_lib  => { OTHERLDFLAGS => $lddflags },
     DEFINE	  => $defs,
     ($Config{osname} eq 'os2' ?
      (
       IMPORTS	  => { xfree => 'emxlibcm.401' }, # Yuck!
      ) : () ),
     INC	  => $RLINC,
     dist	  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz' },
     clean	  => { FILES => "rlver.c rlver$Config{_exe} rlmalloc.c rlmalloc$Config{_exe}" },
);

if ($Config{usesfio} eq 'true') {
    warn <<'EOM';

******************** !!!Warning!!! *********************
** Your Perl is configured as `usesfio' equals true.  **
** Term::ReadLine::Gnu may not work with your Perl.   **
** If it works, let me know your result of `perl -V'. **
********************************************************
EOM
}

exit(0);

########################################################################
# Search a library '$lib' in $Config{libpth} directories, and return
# $lib if exist or undef unless exist.

# ExtUtils::Liblist::ext() do similar job as this subroutine, but it
# warns unnecessary messages.
sub search_lib {
    my ($lib) = @_;
    unless ($lib =~ /^-l/) {
	warn "search_lib: illegal arguments, \`$lib\'.\n";
	return undef;
    }
    my $libbase = 'lib' . substr($lib, 2) . $Config{lib_ext};
    my $libbase_so = 'lib' . substr($lib, 2) . "." . $Config{so};
    foreach (split(' ', $Config{libpth})) {
	if (-f $_ . '/' . $libbase) {
#	    print "$_/$libbase\n";
	    print "Found \`$_/$libbase\'.\n";
	    return $lib;
	} elsif (-f $_ . '/' . $libbase_so) {
#	    print "$_/$libbase_so\n";
	    print "Found \`$_/$libbase_so\'.\n";
	    return $lib;
	}
    }
    return undef;
}

########################################################################
# Check libreadline.a version
#
# Readline 4.2a introduced the macro
#	RL_READLINE_VERSION
#	RL_VERSION_MAJOR
#	RL_VERSION_MINOR
# Someday we don't need this subroutine..
sub check_readline_version {
    my ($RLINC, $RLLIB, $defs, $lddflags, $libs) = @_;
    my $frlver = 'rlver.c';

    # make temp file
    open(F, ">$frlver") || die "Cannot open $frlver:$!\n";
    print F <<'EOF';
/* used by Makefile.pl to check the version of the GNU Readline Library */
#include <stdio.h>
#include <readline/readline.h>
main() { puts(rl_library_version); }
EOF
    close(F);

    # compile it
    my $comp_cmd = "$Config{cc} $RLINC $Config{ccflags} $defs $frlver -o rlver $RLLIB $lddflags $Config{ldflags} $libs";
    print $comp_cmd, "\n";
    system($comp_cmd);
    if ($?) {
	warn <<EOM;
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Could not compile $frlver.

If you have installed the GNU Readline Library (libreadline.{a,so} and
readline/readline.h, etc.) on directories for which your perl is not
configured to search (refer the value of `ccflags' and `libpath' in
the output of `perl -V'), specify the paths as follows;

	perl Makefile.PL --includedir=/yourdir/include --libdir=/yourdir/lib
or
	perl Makefile.PL --prefix=/yourdir

Note that the GNU Readline Library version 2.0 and earlier causes error
here.  Update it to version 2.1 and/or later.

Read INSTALL for more details.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
EOM
    exit $err;
    }

    # execute it and get version
    my $rlver;
    chomp($rlver = `./rlver`);
    # MacOS X 10.4 has libedit by default.
    if ($rlver =~ /EditLine/) {
	warn <<EOM;
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
The libreadline you are using is the libedit library.  Use the GNU Readline Library.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
EOM
	exit $err;
    }
    print "It seems that you have the GNU Readline Library version $rlver.\n";
    # $rlver may be '8.21-beta3' or '4.2a'
    return $rlver =~ /(\d+)\.(\d+)/;
}

########################################################################
# Check whether the xmalloc exported by readline is indeed called
# 'xmalloc' or whether it has been renamed to '_rl_malloc'.
# Contributed by KENTNL at cpan.org
# https://rt.cpan.org/Ticket/Display.html?id=65973
sub guess_malloc_names {
    my ($RLINC, $RLLIB, $defs, $lddflags, $libs) = @_;
    my $frlmalloc = 'rlmalloc.c';

    my @symbol_sets = (
	{
	    # try defaults first
	    # xmalloc  => 'xmalloc',
	    # xrealloc => 'xrealloc',
	    # xfree    => 'xfree',
	},
	{
	    xmalloc  => '_rl_malloc',
	    xrealloc => '_rl_realloc',
	    xfree    => '_rl_free',
	},
    );

    # make temp file
    open(my $F, ">", $frlmalloc) || die "Cannot open $frlmalloc: $!\n";
    print {$F} <<'EOF';
/* used by Makefile.pl to check the names of xmalloc etc. of the GNU Readline Library */
#include <stdio.h>
#include <readline/readline.h>

/* we assume support for ANSI C */
extern void *xmalloc(int);
extern void *xrealloc(void *, int);
extern void xfree(void *);

int main() {
    char *p;
    p = xmalloc(1);
    p[0] = 'o';
    p = xrealloc(p, 3);
    p[1] = 'k';
    p[2] = '\0';
    puts(p);
    xfree(p);
    return 0;
}
EOF
    close($F);

    my $extra_defs;
    for my $symbol_set (@symbol_sets) {
	my $xdef = join " ", map "-D$_=$symbol_set->{$_}", sort keys %$symbol_set;
	# compile it
	my $comp_cmd = "$Config{cc} $RLINC $Config{ccflags} $defs $xdef $frlmalloc -o rlmalloc $RLLIB $lddflags $Config{ldflags} $libs";
	print $comp_cmd, "\n";
	system($comp_cmd);
	unless (system($comp_cmd) || `./rlmalloc` !~ /^ok$/ || $?) {
	    $extra_defs = $xdef;
	    last;
	}
    }

    unless (defined $extra_defs) {
	warn <<EOM;
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
I was unable to find a working xmalloc in your readline library.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
EOM
	exit $err;
    }

    return $extra_defs;
}
# End of Makefile.PL
